001-clarifications-eval.txt

```diff
@@ -0,0 +1,754 @@
+Part 1: Clarifications for Expression Evaluation
+
+In Section 2.5 DWARF Expressions, add the following text after the first
+paragraph:
+
+    --------------------------------------------------------------------
+    [non-normative] The evaluation of a DWARF expression can provide the
+    location of an object, the value of an array bound, the length of a
```

Comment:
This is a bit tricky in my mind because we have two use cases (that we named evaluation kinds), that require either a DWARF expression evaluation or (if I can be so bold to describe it as) a location description expression evaluation. This all depends on the DW_FORM_ of the stream and a use case behind what is represented with that form.

Unfortunately, the DW_FORM_exprloc form has a dual usage depending on the use case and in some instances it is expected to yield a value and in some a location description.

Either way, if the use case is such that the stream should contains a DWARF expression, the evaluation is expected to yield a value, while if the stream contains a location description expression the evaluation result is a location description. In the second case, if the result of the evaluation is a value on the DWARF stack, then the result of an location description expression evaluation is a memory location description.

IMHO here we have an overloaded term where a location description describe the raw byte stream as well as the evaluation result of that stream, where in case of the DWARF expressions we have a separate term for evaluation result called a value.

So maybe the solution here is to have a more general term of just expression that covers both DWARF expressions and the location description expression (name patent pending) so that we don't have to mention those two terms everywhere like it is currently done in the DWARF 5 spec.

Reply:
The underlying issue is that memory location descriptions do not use an operator.  It would be so much clearer if there were DW_OP_mem to turn the result of the preceding DWARF expression into a memory location description.

IMHO we should clarify that an implicit conversion is taking place from an address value computed by a DWARF expression to a memory location description and point out the contexts in which this implicit conversion happens rather than introducing a 'result kind' as part of the clarification.

Reply:
In the 004-locations-on-stack.txt changes, there is text to formally define the implicit conversions.

To me DWARF 5 defines that there are two kinds of DWARF expressions. Those that are defined by a list of operations (termed operation expressions), and those defined by a loclist (termed loclist expressions).

A DWARF operation expression can include operations that produce values and operations that produce location descriptions. For example, the operation expression DW_OP_reg 1 does not produce a value, it produces a register location description. The DW_OP_*piece, DW_OP_stack_value, DW_OP_implicit_pointer, and DW_OP_form_tls_address are examples of other operations that produce location descriptions.

When a DWARF operation expression is evaluated, it is done in a context. Every place that allows DWARF operation expressions, implicitly dictates if a value or location description is required as the result. The changes in 004-locations-on-stack.txt make those implicit rules explicit.

If the context of the evaluation of a DWARF operation expression requires a location description, and its operations actually produce a generic type value, then it is implicitly converted to a memory location description. In 004-locations-on-stack.txt this implicit conversion is formally defined. Given that there is this implicit conversion, there is no need for DWARF 5 to have an operation to do it explicitly. I assume this is preferred as it makes DWARF expressions more compact.

In 004-locations-on-stack.txt DW_OP_addr is defined to produce a location description in the default address space, not a generic type value for the address. In addition, for backwards compatibility, an implicit conversion from a memory location description in the default address space back to a generic type value is defined.

A future proposal (see https://llvm.org/docs/AMDGPUDwarfExtensionsForHeterogeneousDebugging.html#general-support-for-address-spaces) adds DW_OP_form_aspace_address which can create a memory location description in any address space. This is needed for AMD GPU as the compiler needs to specify the address of objects that can be in non-default address space. This could be used to achieve the semantics of your DW_OP_mem by specifying the default address space number (which is defined as 0 for all architectures).

Reply:
I mostly agree with Tony, but I can also appreciate what Markus is saying.

The ambiguity is in the way how the DWARF expression operations are separated from the location description operations, but still can be found in the same byte stream that represent a DW_FORM_exprloc. 

Even in gdb those two kind of operations are treated as one and the same in a sense that there is no error check if a DW_FORM_exprloc byte stream contains location operations when the usage is only DWARF expression evaluation.

Although after 0004, everything falls under the DWARF expression umbrella, it still feels like we need to have an attempt to say how the things are in DWARF 5 so even if that attempt is not completely correct, people will have a chance to understand the problem we are facing.

That being said, this feels like something to be left for the DWARF committee discussions because even here people have a very different views on the topic, so I would be curios to see what the group thinks or if they are even aware of the ambiguity in the first place.

Reply:
I don't believe that this sentence states that the memory location description lives on the DWARF stack. It only states that the evaluation of the DWARF expression can provide an object location among other things.

We could be more precise by saying that:
     "The evaluation of a exprloc can provide the location of an object, the value of an array bound, the length of a...."

and really use the "exprloc" term everywhere where we mean: a byte stream that can hold either DWARF expression or the location description. 

This matches perfectly the existing exprloc definition in the DWARF 5:

                    "This is an unsigned LEB128 length followed by the number of information
                    bytes specified by the length (DW_FORM_exprloc). The information bytes
                    contain a DWARF expression (see Section 2.5 on page 26) or location
                    description (see Section 2.6 on page 38)."

Reply:
>"The evaluation of a exprloc can provide the location of an object, the value of an array bound, the length of a...."

I like that much better.

To the 'result kind' discussion, this would move it to the 'exprloc' form and out of the expression evaluation context.  We will put it back in in chapter 4, of course, but this makes it clear that today that part of the context is a property of the attribute that uses the 'exprloc' form and not a property of the DWARF expression.

Reply:
Well not quite, because the current term DWARF expression evaluation is IMHO wrong and should be called: exprloc evaluation.

Even in GDB's implementation it is in fact the exprloc evaluation and as (I said before) there is no distinction between the two. 

The problem of both DWARF expression and location description being represented with the same byte stream makes the evaluation result kind still belong to the chapter 001, just the term that represents that byte stream should be changed to exprloc from DWARF expression.

Reply:
DWARF 5 defines exprloc as:

> A DWARF expression for a value or a location in the address space of the described program. A leading unsigned LEB128 value (see Section 7.6 on page 221) specifies the number of bytes in the expression.

My reading of this is that any exprloc byte stream is a "DWARF expression" regardless of whether it evaluates to a value or a location description. A DWARF expression is comprised of a stream of operations. Some of those operations create location descriptions and some create values on the stack. Do create a composite or implicit location description you may have operations in the DWARF expression that create both.

I do wonder how precise Chapter 1 has tpo be since the goal is for Chapter 4 to be the precise specification when location descriptions are allowed on the stack. Chapter 1 is just trying to clarify thinks better than they were, not be absolutely complete. I don't think treating DWARF expressions as the union of values and location descriptions is inconsistent with DWARF 5.

Reply:
Unfortunately it is inconsistent.

The DWARF 5 clearly separates those two as two different chapters even, where a location description may or may not use the DWARF expression in their description. 

This is even more obvious when looking at how CFI register rules are defined, where the rules exist in place of location descriptions and then DWARF expressions are used in the same way as in other sections (on their own).

This is very clear in my understanding of the spec and the feeling is that we need to consider that until the chapter 4 where they do become one and the same thing.

Reply:
My interpretation of DWARF 5 is that DWARF expressions cover both DWARF operation expressions (ie a stream of operations) and DWARF loclist expressions.

DWARF loclist expressions can only evaluate to location descriptions.

DWARF operation expressions can evaluate to values or location descriptions.

I agree that DWARF 5 splits the chapters and sections in a way that does not mirror this, but I still think the normative text definitions are consistent with this view.

In our proposal we are making the organization match, and unifying how operation expressions handle evaluating values and location descriptions.

I think trying to clarify DWARF 5 completely in Chapter 1 and 2 is not feasible and is not necessary. @ccoutant has tried to pull out as much clarification that is present in our proposal that can be presented separately in order to reduce the size of Chapter 4. My opinion is that Chapter 1 and 2 should contain improvements over DWARF 5, but do not have to be a complete formalization of DWARF 5 which would be a big undertaking. It is a stepping stone to get to Chapter 4.

Reply:
I guess we would need @ccoutant input here, as a long term member of the DWARF committee.

It seems that all 3 of us have a different interpretation of the DWARF 5 specification regarding the DWARF expression and location description split, so If he also feels that a current wording is good enough to get this part accepted into the issue queue, I don't mind going with it as is.

My only concern is to get the part 1 in a state that allows it to get into the issue queue really :)